src/api/bdaddr.rs

```diff
@@ -0,0 +1,270 @@
+//! Implementation of Bluetooth's MAC address.
+
+use std::fmt;
+use std::str::FromStr;
+use thiserror::Error;
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+#[cfg(feature = "serde")]
+use serde_cr as serde;
+
+use crate::Error;
+
+/// Stores the 6 byte address used to identify Bluetooth devices.
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_cr")
+)]
+#[derive(Copy, Clone, Hash, Eq, PartialEq, Default)]
+#[repr(C)]
+pub struct BDAddr {
+    address: [u8; 6usize],
+}
+
+#[derive(Debug, Error, Clone, PartialEq)]
+pub enum ParseBDAddrError {
+    #[error("Bluetooth address has to be 6 bytes long")]
+    IncorrectByteCount,
+    #[error("All digits in a Bluetooth address must be hex-digits [0-9a-fA-F]")]
+    InvalidDigit,
+}
+
+impl fmt::Display for BDAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        <Self as fmt::LowerHex>::fmt(self, f)
+    }
+}
+
+impl fmt::LowerHex for BDAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let a = &self.address;
+        write!(
+            f,
+            "{:02x}:{:02x}:{:02x}:{:02x}:{:02x}:{:02x}",
+            a[0], a[1], a[2], a[3], a[4], a[5]
+        )
+    }
+}
+
+impl fmt::UpperHex for BDAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let a = &self.address;
+        write!(
+            f,
+            "{:02X}:{:02X}:{:02X}:{:02X}:{:02X}:{:02X}",
+            a[0], a[1], a[2], a[3], a[4], a[5]
+        )
+    }
+}
+
+impl fmt::Debug for BDAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        <Self as fmt::Display>::fmt(self, f)
+    }
+}
+
+impl AsRef<[u8]> for BDAddr {
+    fn as_ref(&self) -> &[u8] {
+        &self.address
+    }
+}
+
+impl From<[u8; 6]> for BDAddr {
+    /// Build an address from an array.
+    ///
+    /// `address[0]` will be the MSB and `address[5]` the LSB.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use btleplug::api::BDAddr;
+    /// let addr: BDAddr = [0x2a, 0xCC, 0x00, 0x34, 0xfa, 0x00].into();
+    /// assert_eq!("2a:cc:00:34:fa:00", addr.to_string());
+    /// ```
+    fn from(address: [u8; 6]) -> Self {
+        Self { address }
+    }
+}
+
+impl<'a> std::convert::TryFrom<&'a [u8]> for BDAddr {
+    type Error = ParseBDAddrError;
+
+    fn try_from(slice: &'a [u8]) -> Result<Self, Self::Error> {
+        if slice.len() < 6 {
+            Err(ParseBDAddrError::IncorrectByteCount)
+        } else {
+            let mut cpy = [0; 6];
+            cpy.copy_from_slice(&slice[..6]);
+            Ok(cpy.into())
+        }
+    }
+}
+
+impl From<u64> for BDAddr {
+    fn from(int: u64) -> Self {
+        let mut cpy = [0; 6];
+        let slice = int.to_be_bytes(); // Reverse order to have MSB on index 0
```

Comment:
Check that the top two bytes are 0.

Reply:
Is this really a problem/error? Would you rather use `TryFrom`?

Reply:
I think that's better. Done this myself.